File: benchmark/synduce/constraints/memo/proper_indexation_sum_lt_pos_v2.f
Inductive List = nil Unit | cons {Int, List};
Inductive IDList = inil Unit | icons {Int, Int, IDList};

length = fix (
  \f: IDList -> Int. \xs: IDList.
  match xs with
    inil _ -> 0
  | icons {_, _, t} -> + 1 (f t)
  end
);

is_indexed = fix (
  \f: IDList -> Bool. \xs: IDList.
  match xs with
    inil _ -> true
  | icons {_, id, t} -> and (f t) (== id (length t))
  end
);

repr = fix (
  \f: IDList -> List. \m: IDList.
  match m with
    inil _ -> nil unit
  | icons {h, _, t} -> cons {h, f t}
  end
);

len = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + 1 (f t)
  end
);

max = \a: Int. \b: Int. if > a b then a else b;

spec = fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 0}
  | cons {h, t} ->
    let res = f t in
      {if > h res.2 then max (+ res.1 h) 0 else res.1, + res.2 1}
  end
);

target = fix (
  \f: IDList -> Compress IDList. \xs: IDList.
  match xs with
    inil _ -> xs
  | icons {h, id, t} -> icons {h, id, f t}
  end
);

main = /*\t: Tree. let mt = gen t in*/
  \m: IDList.
    if is_indexed m then spec (repr (target m)) else {0, 0};

File: ml-benchmark/synduce/constraints/memo/proper_indexation_sum_lt_pos_v2.ml
type idlist = Inil of unit | Icons of int * int * idlist
type lst = Nil of unit | Cons of int * lst

let rec length xs =
  match xs with
  | Inil _ -> 0
  | Icons (_, _, t) -> 1 + length t

let rec is_indexed xs =
  match xs with
  | Inil _ -> true
  | Icons (_, id, t) -> is_indexed t && id = length t

let rec repr m =
  match m with
  | Inil _ -> Nil ()
  | Icons (h, _, t) -> Cons (h, repr t)

let rec len xs =
  match xs with
  | Nil _ -> 0
  | Cons (_, t) -> 1 + len t

let max a b =
  if a > b then a else b

let rec spec xs =
  match xs with
  | Nil _ -> (0, 0)
  | Cons (h, t) ->
      let res = spec t in
      match res with
      | (r1, r2) ->
          let a =
            if h > r2 then
              max (r1 + h) 0
            else
              r1
          in
          let b = r2 + 1 in
          (a, b)

let rec target xs =
  match xs with
  | Inil _ -> xs
  | Icons (h, id, t) -> Icons (h, id, target t)

let program m =
  if is_indexed m then spec (repr (target m)) else (0, 0)